src/library/scala/collection/mutable/HashTable.scala

```diff
@@ -357,14 +360,14 @@ trait HashTable[A, Entry >: Null <: HashEntry[A, Entry]] extends HashTable.HashU
 
   protected def elemEquals(key1: A, key2: A): Boolean = (key1 == key2)
 
-  // Note:
-  // we take the most significant bits of the hashcode, not the lower ones
-  // this is of crucial importance when populating the table in parallel
-  protected final def index(hcode: Int) = {
+  /**
+    * Note: we take the most significant bits of the hashcode, not the lower ones
+    * this is of crucial importance when populating the table in parallel
+    */
+  protected final def index(hcode: Int): Int = {
     val ones = table.length - 1
-    val improved = improve(hcode, seedvalue)
-    val shifted = (improved >> (32 - java.lang.Integer.bitCount(ones))) & ones
-    shifted
+    val exponent = Integer.numberOfLeadingZeros(ones)
+    (improve(hcode, seedvalue) >>> exponent) & ones
```

Comment:
Now I realize: since `ones` is of the form `00...0011...11`, I do not think there is any value of `x` for which `(x >>> exponent) & ones` is different from `(x >>> exponent)`. It seems to me that `& ones` is redundant and can be dropped. That's because `& ones` masks off the `exponent` most significant bits of `(x >>> exponent)`, which by construction already has `0`'s in all its `exponent` most significant bits.